### PR TITLE
Handle infinitesimal and negative variance in `UpdateBy#RollingStd()`

### LINKED ISSUE #4397: `rolling_std_tick` behaves incorrectly on `float` columns
**Description**

When `rolling_std_tick` gets applied to a column of type `float` where all of the entries are the same, the behavior is unpredictable and often incorrect.

**Steps to reproduce**
Run the following code in the Deephaven IDE:
```
from deephaven import empty_table
from deephaven.updateby import rolling_std_tick

test = empty_table(100).update("X = 1.2345")
result1 = test.update_by(ops = rolling_std_tick("X", rev_ticks = 2))
result2 = test.update_by(ops = rolling_std_tick("X", rev_ticks = 3))
result3 = test.update_by(ops = rolling_std_tick("X", rev_ticks = 4))
result4 = test.update_by(ops = rolling_std_tick("X", rev_ticks = 5))
result5 = test.update_by(ops = rolling_std_tick("X", rev_ticks = 6))
```

**Expected results**

I expect every `result*` table to contain a single column with a single `NaN` at the beginning and only `0.0000` after that.

**Actual results**

Instead, all of `result1` through `result5` have different results, with strange mixtures of `NaN`, `0.0000`, and `0.00000001490116`. 

**Versions**
 - Deephaven: 0.28.0.
- OS: MacOS 13.2.1
- Browser: Safari 16.3

### PR DESCRIPTION
**Pull request by lbooker42**
When computing `std` / `variance` on effectively constant values, floating point error can result in the computed `variance` becoming a negative value, forcing `std` (which is `sqrt(variance)`) to become `NaN` in the output columns.   This PR clamps variance values to zero when smaller than a defined threshold (currently 1e-12) which also corrects the case when mathematically the variance should be zero but is instead a near-zero floating point value.

Will close #4397

### CODE REVIEW

**Review by nbauernfeind — DISMISSED** [review]
From a code perspective, this LGTM.

**Review by alexpeters1208 — APPROVED** [review]
Tested from the R client, everything behaves as expected.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingstd/ByteRollingStdOperator.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 public class ByteRollingStdOperator extends BaseDoubleUpdateByOperator {
     private static final int BUFFER_INITIAL_CAPACITY = 128;
+    private static final double VARIANCE_THRESHOLD = 1e-12;
```
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingstd/ByteRollingStdOperator.java:25]: As presented, this change can not go in.  For example, consider a physics group that is using DH to measure the mass of an electron (9.11E-28 g), which I will call m_e.  The value of m_e is >> 2E-308.  As a result, it is very reasonable to compute the variance of m_e measurements and expect good numerics.  By choosing a hard limit of 1E-12, the m_e experiment always returns a variance of 0.0, which is BS.

It is very reasonable to put a hard limit of 0.0, but putting a non-zero constant threshold is not viable.

Numpy's impl is here.  I have not dug through all of the details, but you can see that there is no hard limit.
https://github.com/numpy/numpy/blob/v1.25.0/numpy/core/fromnumeric.py#L3654-L3788
https://github.com/numpy/numpy/blob/v1.25.0/numpy/core/_methods.py#L135

We may be able to look at the constituent values to determine something useful.  e.g. if all the values are the same, the variance is zero.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingstd/ByteRollingStdOperator.java (hunk 1 above)
Comment (lbooker42) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingstd/ByteRollingStdOperator.java]: Implemented the floor of `0.0` to cover only the negative variance case.  This should eliminate the `NaN` while still giving accurate results (to the limits of floating point precision).

**Review by chipkent — DISMISSED** [review]
Approved pending @alexpeters1208 testing.

**Comment by alexpeters1208** [comment]
> Approved pending @alexpeters1208 testing.

Here is some R output:

```
Expected (rev_ticks = 2):
   col1   col2      col3
1    NA hello!        NA
2     0 hello! 0.5790145
3     0 hello! 1.2703186
4     0 hello! 1.2680206
5     0 hello! 1.5112402
6     0 hello! 1.9807812
7     0 hello! 1.6991713
8     0 hello! 0.4118309
9     0 hello! 0.7772836
10    0 hello! 0.6823800
11    0 hello! 0.2914419
12    0 hello! 0.1352065
13    0 hello! 0.5665835
14    0 hello! 0.1108007
15    0 hello! 0.5551575

Actual (rev_ticks = 2):
   col1   col2      col3
1   NaN hello!       NaN
2     0 hello! 0.5790145
3     0 hello! 1.2703186
4     0 hello! 1.2680206
5     0 hello! 1.5112402
6     0 hello! 1.9807812
7     0 hello! 1.6991713
8     0 hello! 0.4118309
9     0 hello! 0.7772836
10    0 hello! 0.6823800
11    0 hello! 0.2914419
12    0 hello! 0.1352065
13    0 hello! 0.5665835
14    0 hello! 0.1108007
15    0 hello! 0.5551575

=================================================================================

Expected (rev_ticks = 5):
   col1   col2      col3
1    NA hello!        NA
2     0 hello! 0.5790145
3     0 hello! 1.3377843
4     0 hello! 1.1041442
5     0 hello! 1.2777865
6     0 hello! 1.2343740
7     0 hello! 1.2487559
8     0 hello! 1.1782527
9     0 hello! 1.1678631
10    0 hello! 0.9563309
11    0 hello! 0.6279583
12    0 hello! 0.4299351
13    0 hello! 0.4532259
14    0 hello! 0.5115861
15    0 hello! 0.4255851

Actual (rev_ticks = 5):
           col1   col2      col3
1           NaN hello!       NaN
2  0.000000e+00 hello! 0.5790145
3  0.000000e+00 hello! 1.3377843
4  0.000000e+00 hello! 1.1041442
5  0.000000e+00 hello! 1.2777865
6  4.214685e-08 hello! 1.2343740
7  4.214685e-08 hello! 1.2487559
8  0.000000e+00 hello! 1.1782527
9  0.000000e+00 hello! 1.1678631
10 4.214685e-08 hello! 0.9563309
11 4.214685e-08 hello! 0.6279583
12 0.000000e+00 hello! 0.4299351
13 0.000000e+00 hello! 0.4532259
14 4.214685e-08 hello! 0.5115861
15 4.214685e-08 hello! 0.4255851

=================================================================================

Expected (rev_ticks = 10):
   col1   col2      col3
1    NA hello!        NA
2     0 hello! 0.5790145
3     0 hello! 1.3377843
4     0 hello! 1.1041442
5     0 hello! 1.2777865
6     0 hello! 1.2575446
7     0 hello! 1.2535357
8     0 hello! 1.1772800
9     0 hello! 1.1198454
10    0 hello! 1.0642406
11    0 hello! 0.9337404
12    0 hello! 0.8920184
13    0 hello! 0.8645640
14    0 hello! 0.8796553
15    0 hello! 0.7138739

Actual (rev_ticks = 10):
           col1   col2      col3
1           NaN hello!       NaN
2  0.000000e+00 hello! 0.5790145
3  0.000000e+00 hello! 1.3377843
4  0.000000e+00 hello! 1.1041442
5  0.000000e+00 hello! 1.2777865
6  0.000000e+00 hello! 1.2575446
7  0.000000e+00 hello! 1.2535357
8  0.000000e+00 hello! 1.1772800
9  0.000000e+00 hello! 1.1198454
10 0.000000e+00 hello! 1.0642406
11 4.214685e-08 hello! 0.9337404
12 4.214685e-08 hello! 0.8920184
13 0.000000e+00 hello! 0.8645640
14 4.214685e-08 hello! 0.8796553
15 4.214685e-08 hello! 0.7138739
```

**Comment by alexpeters1208** [comment]
> > Approved pending @alexpeters1208 testing.
> 
> Here is some R output:
> 
> ```
> Expected (rev_ticks = 2):
>    col1   col2      col3
> 1    NA hello!        NA
> 2     0 hello! 0.5790145
> 3     0 hello! 1.2703186
> 4     0 hello! 1.2680206
> 5     0 hello! 1.5112402
> 6     0 hello! 1.9807812
> 7     0 hello! 1.6991713
> 8     0 hello! 0.4118309
> 9     0 hello! 0.7772836
> 10    0 hello! 0.6823800
> 11    0 hello! 0.2914419
> 12    0 hello! 0.1352065
> 13    0 hello! 0.5665835
> 14    0 hello! 0.1108007
> 15    0 hello! 0.5551575
> 
> Actual (rev_ticks = 2):
>    col1   col2      col3
> 1   NaN hello!       NaN
> 2     0 hello! 0.5790145
> 3     0 hello! 1.2703186
> 4     0 hello! 1.2680206
> 5     0 hello! 1.5112402
> 6     0 hello! 1.9807812
> 7     0 hello! 1.6991713
> 8     0 hello! 0.4118309
> 9     0 hello! 0.7772836
> 10    0 hello! 0.6823800
> 11    0 hello! 0.2914419
> 12    0 hello! 0.1352065
> 13    0 hello! 0.5665835
> 14    0 hello! 0.1108007
> 15    0 hello! 0.5551575
> 
> =================================================================================
> 
> Expected (rev_ticks = 5):
>    col1   col2      col3
> 1    NA hello!        NA
> 2     0 hello! 0.5790145
> 3     0 hello! 1.3377843
> 4     0 hello! 1.1041442
> 5     0 hello! 1.2777865
> 6     0 hello! 1.2343740
> 7     0 hello! 1.2487559
> 8     0 hello! 1.1782527
> 9     0 hello! 1.1678631
> 10    0 hello! 0.9563309
> 11    0 hello! 0.6279583
> 12    0 hello! 0.4299351
> 13    0 hello! 0.4532259
> 14    0 hello! 0.5115861
> 15    0 hello! 0.4255851
> 
> Actual (rev_ticks = 5):
>            col1   col2      col3
> 1           NaN hello!       NaN
> 2  0.000000e+00 hello! 0.5790145
> 3  0.000000e+00 hello! 1.3377843
> 4  0.000000e+00 hello! 1.1041442
> 5  0.000000e+00 hello! 1.2777865
> 6  4.214685e-08 hello! 1.2343740
> 7  4.214685e-08 hello! 1.2487559
> 8  0.000000e+00 hello! 1.1782527
> 9  0.000000e+00 hello! 1.1678631
> 10 4.214685e-08 hello! 0.9563309
> 11 4.214685e-08 hello! 0.6279583
> 12 0.000000e+00 hello! 0.4299351
> 13 0.000000e+00 hello! 0.4532259
> 14 4.214685e-08 hello! 0.5115861
> 15 4.214685e-08 hello! 0.4255851
> 
> =================================================================================
> 
> Expected (rev_ticks = 10):
>    col1   col2      col3
> 1    NA hello!        NA
> 2     0 hello! 0.5790145
> 3     0 hello! 1.3377843
> 4     0 hello! 1.1041442
> 5     0 hello! 1.2777865
> 6     0 hello! 1.2575446
> 7     0 hello! 1.2535357
> 8     0 hello! 1.1772800
> 9     0 hello! 1.1198454
> 10    0 hello! 1.0642406
> 11    0 hello! 0.9337404
> 12    0 hello! 0.8920184
> 13    0 hello! 0.8645640
> 14    0 hello! 0.8796553
> 15    0 hello! 0.7138739
> 
> Actual (rev_ticks = 10):
>            col1   col2      col3
> 1           NaN hello!       NaN
> 2  0.000000e+00 hello! 0.5790145
> 3  0.000000e+00 hello! 1.3377843
> 4  0.000000e+00 hello! 1.1041442
> 5  0.000000e+00 hello! 1.2777865
> 6  0.000000e+00 hello! 1.2575446
> 7  0.000000e+00 hello! 1.2535357
> 8  0.000000e+00 hello! 1.1772800
> 9  0.000000e+00 hello! 1.1198454
> 10 0.000000e+00 hello! 1.0642406
> 11 4.214685e-08 hello! 0.9337404
> 12 4.214685e-08 hello! 0.8920184
> 13 0.000000e+00 hello! 0.8645640
> 14 4.214685e-08 hello! 0.8796553
> 15 4.214685e-08 hello! 0.7138739
> ```

I can do an `is.close()` to get the tests to pass, if that's what we want to do. Only need to decide whether this kind of systematic error is acceptable

**Comment by chipkent** [comment]
If you look at the code, you see:
image

The results indicate that `valueSum` or `valueSquareSum` is not zero, when all inputs are zero.  This makes me think that the error is not in these equations but is in what computes these values.

**Comment by alexpeters1208** [comment]
Input table (truncated):
```
   col1   col2        col3
1  3.14 hello!  0.39981343
2  3.14 hello! -1.09386350
3  3.14 hello!  0.07419756
4  3.14 hello! -2.80561783
5  3.14 hello!  0.87295074
6  3.14 hello!  0.64722657
7  3.14 hello!  1.00876972
8  3.14 hello! -0.32908439
9  3.14 hello!  0.70285781
10 3.14 hello! -2.34024862
11 3.14 hello!  1.19070406
12 3.14 hello! -0.24047068
13 3.14 hello!  1.17721254
14 3.14 hello! -0.27264104
15 3.14 hello!  1.67086058
```

Metadata as seen by the Deephaven IDE:
```
  Name         DataType ColumnType IsPartitioning IsGrouping
1 col1           double     Normal          FALSE      FALSE
2 col2 java.lang.String     Normal          FALSE      FALSE
3 col3           double     Normal          FALSE      FALSE
```

**Comment by chipkent** [comment]
Here is some python test code that illustrates the numerics:
```
def comp_var(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return v2s / (n-1) - vs*vs/n/(n-1)


for i in range(2, 21):
    print(f"{i} {comp_var(3.14,i)}")
```

image

**Comment by chipkent** [comment]
If the example is then rearranged to compute the variance in two different ways, there are two different sets of rounding errors:
```
def comp_var(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return v2s / (n-1) - vs*vs/n/(n-1)

def comp_var2(x:float, n:int) -> float:
    vs = x*n
    xbar = vs/n
    return ((x-xbar)*n)/n


for i in range(2, 21):
    print(f"{i} {comp_var(3.14,i)} {comp_var2(3.14,i)}")
```
image

Comp_var2 has lower rounding errors, but it requires much much larger state for the rolling calculation.

**Comment by chipkent** [comment]
Rearranging the terms leaves a result that seems to have better rounding properties.  See `comp_var3`.
```

def comp_var(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return v2s / (n-1) - vs*vs/n/(n-1)

def comp_var2(x:float, n:int) -> float:
    vs = x*n
    xbar = vs/n
    return ((x-xbar)*n)/n

def comp_var3(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return (v2s - vs*(vs/n))/(n-1)

for i in range(2, 21):
    print(f"{i} {comp_var(3.14,i)} {comp_var2(3.14,i)} {comp_var3(3.14,i)}")
```

image

**Comment by chipkent** [comment]
The actual value used will see some fluctuation in the tolerance of the variance.  See:
```

def comp_var(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return v2s / (n-1) - vs*vs/n/(n-1)

def comp_var2(x:float, n:int) -> float:
    vs = x*n
    xbar = vs/n
    return ((x-xbar)*n)/n

def comp_var3(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return (v2s - vs*(vs/n))/(n-1)

x_test = 93.24
for i in range(2, 21):
    print(f"{i} {comp_var(x_test,i)} {comp_var2(x_test,i)} {comp_var3(x_test,i)}")
```
image

**Comment by chipkent** [comment]
Here is a new alternative tha may be more robust to IEEE754.

```

import numpy as np

def comp_var(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return v2s / (n-1) - vs*vs/n/(n-1)

def comp_var2(x:float, n:int) -> float:
    vs = x*n
    xbar = vs/n
    return ((x-xbar)*n)/n

def comp_var3(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    return (v2s - vs*(vs/n))/(n-1)

def comp_var4(x:float, n:int) -> float:
    vs = x*n
    v2s = x*x*n
    vs2bar = vs*(vs/n)
    delta = v2s-vs2bar
    eps = np.spacing(v2s)
    rel_eps = delta/eps
    return delta/(n-1) if np.abs(rel_eps) > 1 else 0.0 

x_test = 111.2
for i in range(2, 21):
    print(f"{i} {comp_var(x_test,i)} {comp_var2(x_test,i)} {comp_var3(x_test,i)} {comp_var4(x_test,i)}")
```

image

**Comment by lbooker42** [comment]
Adopted suggestion 4 and applied it to `update_by#rolling_std` and to `Numeric.ftl` variance computation.

Variance and standard deviation should now produce identical results whether called as a vector function or as part of `update_by`.

**Review by chipkent — APPROVED** [review]

**Pull request merged**